File: go/sorting.go
package main

import (
	"bufio"
	"os"
	"strconv"
)

func LoadDataset(path string) []int64 {
	file, err := os.Open(path)
	if err != nil {
		panic(err)
	}
	defer file.Close()

	var data []int64
	scanner := bufio.NewScanner(file)

	for scanner.Scan() {
		v, err := strconv.ParseInt(scanner.Text(), 10, 64)
		if err != nil {
			panic(err)
		}
		data = append(data, v)
	}

	return data
}

func checksum(data []int64) uint64 {
	var sum uint64 = 0

	for _, v := range data {
		sum += uint64(v)
	}

	return sum
}

func RunBubble(data []int64) uint64 {
	n := len(data)

	for i := range n {
		for j := 0; j < n-i-1; j++ {
			if data[j] > data[j+1] {
				data[j], data[j+1] = data[j+1], data[j]
			}
		}
	}

	return checksum(data)
}

func RunQuick(data []int64) uint64 {
	quicksort(data, 0, len(data)-1)
	return checksum(data)
}

func quicksort(arr []int64, low, high int) {
	if low < high {
		p := partition(arr, low, high)
		quicksort(arr, low, p-1)
		quicksort(arr, p+1, high)
	}
}

func partition(arr []int64, low, high int) int {
	pivot := arr[high]
	i := low

	for j := low; j < high; j++ {
		if arr[j] < pivot {
			arr[i], arr[j] = arr[j], arr[i]
			i++
		}
	}

	arr[i], arr[high] = arr[high], arr[i]
	return i
}

func RunMerge(data []int64) uint64 {
	mergeSort(data)
	return checksum(data)
}

func mergeSort(arr []int64) {
	if len(arr) <= 1 {
		return
	}

	mid := len(arr) / 2
	left := make([]int64, mid)
	right := make([]int64, len(arr)-mid)

	copy(left, arr[:mid])
	copy(right, arr[mid:])

	mergeSort(left)
	mergeSort(right)

	merge(arr, left, right)
}

func merge(result, left, right []int64) {
	i, j, k := 0, 0, 0

	for i < len(left) && j < len(right) {
		if left[i] <= right[j] {
			result[k] = left[i]
			i++
		} else {
			result[k] = right[j]
			j++
		}
		k++
	}

	for i < len(left) {
		result[k] = left[i]
		i++
		k++
	}

	for j < len(right) {
		result[k] = right[j]
		j++
		k++
	}
}

File: rust/src/sorting.rs
use std::fs::File;
use std::io::{BufRead, BufReader};

// Helper
pub fn load_dataset(path: String) -> Vec<i64> {
    let file = File::open(path).expect("Unable to open dataset file");
    let reader = BufReader::new(file);

    reader
        .lines()
        .map(|line| {
            line.expect("Invalid line")
                .parse::<i64>()
                .expect("Invalid number")
        })
        .collect()
}

fn checksum(data: &[i64]) -> u64 {
    data.iter().fold(0u64, |acc, &x| acc.wrapping_add(x as u64))
}

// Bubblesort
pub fn run_bubble(mut data: Vec<i64>) -> u64 {
    let n = data.len();
    for i in 0..n {
        for j in 0..(n - i - 1) {
            if data[j] > data[j + 1] {
                data.swap(j, j + 1);
            }
        }
    }
    checksum(&data)
}

// Quicksort
pub fn run_quick(mut data: Vec<i64>) -> u64 {
    quicksort(&mut data);
    checksum(&data)
}

fn quicksort(arr: &mut [i64]) {
    if arr.len() <= 1 {
        return;
    }

    let pivot_index = partition(arr);
    let (left, right) = arr.split_at_mut(pivot_index);
    quicksort(left);
    quicksort(&mut right[1..]);
}

fn partition(arr: &mut [i64]) -> usize {
    let len = arr.len();
    let pivot = arr[len - 1];
    let mut i = 0;

    for j in 0..len - 1 {
        if arr[j] < pivot {
            arr.swap(i, j);
            i += 1;
        }
    }

    arr.swap(i, len - 1);
    i
}

// Mergesort
pub fn run_merge(mut data: Vec<i64>) -> u64 {
    merge_sort(&mut data);
    checksum(&data)
}

fn merge_sort(arr: &mut [i64]) {
    let len = arr.len();
    if len <= 1 {
        return;
    }

    let mid = len / 2;
    merge_sort(&mut arr[..mid]);
    merge_sort(&mut arr[mid..]);

    let mut merged = arr.to_vec();
    merge(&arr[..mid], &arr[mid..], &mut merged[..]);
    arr.copy_from_slice(&merged);
}

fn merge(left: &[i64], right: &[i64], result: &mut [i64]) {
    let mut i = 0;
    let mut j = 0;
    let mut k = 0;

    while i < left.len() && j < right.len() {
        if left[i] <= right[j] {
            result[k] = left[i];
            i += 1;
        } else {
            result[k] = right[j];
            j += 1;
        }
        k += 1;
    }

    while i < left.len() {
        result[k] = left[i];
        i += 1;
        k += 1;
    }

    while j < right.len() {
        result[k] = right[j];
        j += 1;
        k += 1;
    }
}
